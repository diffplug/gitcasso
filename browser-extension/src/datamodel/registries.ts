import type { OverType } from '../overtype/mock-overtype'
import type { CommentEnhancer, CommentSpot } from './enhancer'
import { GitHubHandler as GitHubEnhancer } from './handlers/github-handler'
import { RedditHandler as RedditEnhancer } from './handlers/reddit-handler'

export interface EnhancedTextarea<T extends CommentSpot = CommentSpot> {
  element: HTMLTextAreaElement
  context: T
  handler: CommentEnhancer<T>
  overtype: OverType
}

export class EnhancerRegistry {
  private enhancers = new Set<CommentEnhancer<any>>()

  constructor() {
    // Register all available handlers
    this.register(new GitHubEnhancer())
    this.register(new RedditEnhancer())
  }

  private register<T extends CommentSpot>(handler: CommentEnhancer<T>): void {
    this.enhancers.add(handler)
  }

  getHandlerForType(type: string): CommentEnhancer<any> | null {
    for (const handler of this.enhancers) {
      if (handler.forCommentTypes().includes(type)) {
        return handler
      }
    }
    return null
  }

  identifyTextarea(textarea: HTMLTextAreaElement): EnhancedTextarea<any> | null {
    for (const handler of this.enhancers) {
      try {
        const result = handler.tryToEnhance(textarea)
        if (result) {
          const [overtype, context] = result
          return { context, element: textarea, handler, overtype }
        }
      } catch (error) {
        console.warn('Handler failed to identify textarea:', error)
      }
    }
    return null
  }

  getAllHandlers(): CommentEnhancer<any>[] {
    return Array.from(this.enhancers)
  }

  getCommentTypesForHandler(handler: CommentEnhancer<any>): string[] {
    return handler.forCommentTypes()
  }
}

export class TextareaRegistry {
  private textareas = new Map<HTMLTextAreaElement, EnhancedTextarea<any>>()

  register<T extends CommentSpot>(textareaInfo: EnhancedTextarea<T>): void {
    this.textareas.set(textareaInfo.element, textareaInfo)
    // TODO: register as a draft in progress with the global list
  }

  unregisterDueToModification(textarea: HTMLTextAreaElement): void {
    if (this.textareas.has(textarea)) {
      // TODO: register as abandoned or maybe submitted with the global list
      this.textareas.delete(textarea)
    }
  }

  get(textarea: HTMLTextAreaElement): EnhancedTextarea<any> | undefined {
    return this.textareas.get(textarea)
  }
}
