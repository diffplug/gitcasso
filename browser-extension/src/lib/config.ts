const MODES = ['PROD', 'PLAYGROUNDS_PR'] as const

export type ModeType = (typeof MODES)[number]

export const CONFIG = {
  ADDED_OVERTYPE_CLASS: 'gitcasso-overtype',
  DEBUG: true, // enabled debug logging
  EXTENSION_NAME: 'gitcasso', // decorates logs
  MODE: 'PLAYGROUNDS_PR' satisfies ModeType,
} as const
