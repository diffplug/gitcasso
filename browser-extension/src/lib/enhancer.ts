import type { OverTypeInstance } from '../overtype/overtype'

/**
 * stores enough info about the location of a draft to:
 * - display it in a table
 * - reopen the draft in-context
 */
export interface CommentSpot {
  unique_key: string
  type: string
}

/** wraps the textareas of a given platform with Gitcasso's enhancements */
export interface CommentEnhancer<Spot extends CommentSpot = CommentSpot> {
  /** guarantees to only return a type within this list */
  forSpotTypes(): string[]
  /**
   * whenever a new `textarea` is added to any webpage, this method is called.
   * if we return non-null, then we become the handler for that text area.
   */
  tryToEnhance(textarea: HTMLTextAreaElement): [OverTypeInstance, Spot] | null

  tableIcon(spot: Spot): string
  tableTitle(spot: Spot): string
  buildUrl(spot: Spot): string
}
