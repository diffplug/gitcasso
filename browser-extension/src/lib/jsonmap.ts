/** Map which can take any JSON as key and uses its sorted serialization as the underlying key. */
export class JsonMap<K, V> {
  private map = new Map<string, V>()

  set(key: K, value: V): this {
    this.map.set(stableStringify(key), value)
    return this
  }

  get(key: K): V | undefined {
    return this.map.get(stableStringify(key))
  }

  has(key: K): boolean {
    return this.map.has(stableStringify(key))
  }

  delete(key: K): boolean {
    return this.map.delete(stableStringify(key))
  }

  clear(): void {
    this.map.clear()
  }

  get size(): number {
    return this.map.size
  }

  *values(): IterableIterator<V> {
    yield* this.map.values()
  }

  *entries(): IterableIterator<[K, V]> {
    for (const [stringKey, value] of this.map.entries()) {
      yield [JSON.parse(stringKey) as K, value]
    }
  }

  *keys(): IterableIterator<K> {
    for (const stringKey of this.map.keys()) {
      yield JSON.parse(stringKey) as K
    }
  }

  [Symbol.iterator](): IterableIterator<[K, V]> {
    return this.entries()
  }
}

function stableStringify(v: unknown): string {
  return JSON.stringify(v, (_k, val) => {
    if (val && typeof val === 'object' && !Array.isArray(val)) {
      // sort object keys recursively
      return Object.keys(val as Record<string, unknown>)
        .sort()
        .reduce(
          (acc, k) => {
            ;(acc as any)[k] = (val as any)[k]
            return acc
          },
          {} as Record<string, unknown>,
        )
    }
    return val
  })
}
