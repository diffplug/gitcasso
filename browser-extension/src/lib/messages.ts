import type { CommentDraft, CommentEvent, CommentSpot } from './enhancer'

// Content -> Background messages (already well-typed as CommentEvent)
export type ContentToBackgroundMessage = CommentEvent

// Popup -> Background messages
export interface GetOpenSpotsMessage {
  type: 'GET_OPEN_SPOTS'
}

export interface SwitchToTabMessage {
  type: 'SWITCH_TO_TAB'
  tabId: number
  windowId: number
}

export type PopupToBackgroundMessage = GetOpenSpotsMessage | SwitchToTabMessage

// All messages sent to background
export type ToBackgroundMessage = ContentToBackgroundMessage | PopupToBackgroundMessage

// Background -> Popup responses
export interface GetOpenSpotsResponse {
  spots: Array<{
    tab: {
      tabId: number
      windowId: number
    }
    spot: CommentSpot
    drafts: Array<[number, CommentDraft]>
  }>
}

// Type guard functions
export function isContentToBackgroundMessage(message: any): message is ContentToBackgroundMessage {
  return (
    message &&
    typeof message.type === 'string' &&
    (message.type === 'ENHANCED' || message.type === 'DESTROYED') &&
    message.spot
  )
}

export function isPopupToBackgroundMessage(message: any): message is PopupToBackgroundMessage {
  return (
    message &&
    typeof message.type === 'string' &&
    (message.type === 'GET_OPEN_SPOTS' || message.type === 'SWITCH_TO_TAB')
  )
}

export function isGetOpenSpotsMessage(message: any): message is GetOpenSpotsMessage {
  return message && message.type === 'GET_OPEN_SPOTS'
}

export function isSwitchToTabMessage(message: any): message is SwitchToTabMessage {
  return (
    message &&
    message.type === 'SWITCH_TO_TAB' &&
    typeof message.tabId === 'number' &&
    typeof message.windowId === 'number'
  )
}

// Message handler types
export type BackgroundMessageHandler = (
  message: ToBackgroundMessage,
  sender: any,
  sendResponse: (response?: any) => void,
) => boolean | undefined

export type PopupMessageSender = {
  sendMessage<T extends PopupToBackgroundMessage>(
    message: T,
  ): Promise<T extends GetOpenSpotsMessage ? GetOpenSpotsResponse : void>
}
