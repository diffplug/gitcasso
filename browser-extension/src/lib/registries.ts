import type { OverType } from '../overtype/mock-overtype'
import type { CommentEnhancer, CommentSpot } from './enhancer'
import { GitHubAddCommentEnhancer } from './enhancers/github'

export interface EnhancedTextarea<T extends CommentSpot = CommentSpot> {
  textarea: HTMLTextAreaElement
  spot: T
  handler: CommentEnhancer<T>
  overtype: OverType
}

export class EnhancerRegistry {
  private enhancers = new Set<CommentEnhancer<any>>()

  constructor() {
    // Register all available handlers
    this.register(new GitHubAddCommentEnhancer())
  }

  private register<T extends CommentSpot>(handler: CommentEnhancer<T>): void {
    this.enhancers.add(handler)
  }

  tryToEnhance(textarea: HTMLTextAreaElement): EnhancedTextarea<any> | null {
    for (const handler of this.enhancers) {
      try {
        const result = handler.tryToEnhance(textarea)
        if (result) {
          const [overtype, spot] = result
          return { handler, overtype, spot, textarea }
        }
      } catch (error) {
        console.warn('Handler failed to identify textarea:', error)
      }
    }
    return null
  }

  getEnhancerCount(): number {
    return this.enhancers.size
  }
}

export class TextareaRegistry {
  private textareas = new Map<HTMLTextAreaElement, EnhancedTextarea<any>>()

  register<T extends CommentSpot>(textareaInfo: EnhancedTextarea<T>): void {
    this.textareas.set(textareaInfo.textarea, textareaInfo)
    // TODO: register as a draft in progress with the global list
  }

  unregisterDueToModification(textarea: HTMLTextAreaElement): void {
    if (this.textareas.has(textarea)) {
      // TODO: register as abandoned or maybe submitted with the global list
      this.textareas.delete(textarea)
    }
  }

  get(textarea: HTMLTextAreaElement): EnhancedTextarea<any> | undefined {
    return this.textareas.get(textarea)
  }
}
