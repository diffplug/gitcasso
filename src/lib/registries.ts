import type { OverTypeInstance } from 'overtype'
import OverType from 'overtype'
import type {
  CommentEnhancer,
  CommentEvent,
  CommentEventType,
  CommentSpot,
  StrippedLocation,
} from './enhancer'
import { CommentEnhancerMissing } from './enhancers/CommentEnhancerMissing'
import { GitHubEditEnhancer } from './enhancers/github/GitHubEditEnhancer'
import { GitHubIssueAppendEnhancer } from './enhancers/github/GitHubIssueAppendEnhancer'
import { GitHubIssueCreateEnhancer } from './enhancers/github/GitHubIssueCreateEnhancer'
import { GitHubPrAppendEnhancer } from './enhancers/github/GitHubPrAppendEnhancer'
import { GitHubPrCreateEnhancer } from './enhancers/github/GitHubPrCreateEnhancer'

export interface EnhancedTextarea<T extends CommentSpot = CommentSpot> {
  textarea: HTMLTextAreaElement
  spot: T
  enhancer: CommentEnhancer<T>
  overtype: OverTypeInstance
  cleanup?: () => void
}

export class EnhancerRegistry {
  private enhancers = new Set<CommentEnhancer>()
  byType = new Map<string, CommentEnhancer>()

  constructor() {
    // Register all available handlers
    this.register(new GitHubEditEnhancer())
    this.register(new GitHubIssueAppendEnhancer())
    this.register(new GitHubIssueCreateEnhancer())
    this.register(new GitHubPrAppendEnhancer())
    this.register(new GitHubPrCreateEnhancer())
    const textColor = 'rgb(31, 35, 40)'
    const headingColor = 'rgb(174, 52, 151)'
    OverType.setTheme({
      colors: {
        blockquote: 'rgb(89, 99, 110)',
        code: '#59636e',
        codeBg: '#f6f8fa',
        cursor: '#000000',
        em: 'rgb(126, 123, 255)',
        h1: headingColor,
        h2: headingColor,
        h3: headingColor,
        hr: '#5a7a9b',
        link: 'rgb(9, 105, 218)',
        selection: 'rgba(0, 123, 255, 0.3)',
        strong: 'rgb(45, 1, 142)',
        syntaxMarker: textColor,
        text: textColor,
      },
      name: 'custom-github',
    })
  }

  private register<T extends CommentSpot>(enhancer: CommentEnhancer<T>): void {
    this.enhancers.add(enhancer)
    for (const spotType of enhancer.forSpotTypes()) {
      this.byType.set(spotType, enhancer)
    }
  }

  enhancerFor<T extends CommentSpot>(spot: T): CommentEnhancer<T> {
    return (this.byType.get(spot.type) || new CommentEnhancerMissing()) as CommentEnhancer<T>
  }

  tryToEnhance(textarea: HTMLTextAreaElement, location: StrippedLocation): EnhancedTextarea | null {
    for (const enhancer of this.enhancers) {
      try {
        const spot = enhancer.tryToEnhance(textarea, location)
        if (spot) {
          const { instance: overtype, cleanup } = enhancer.enhance(textarea, spot)
          this.handleDelayedValueInjection(overtype)
          return { enhancer, overtype, spot, textarea, ...(cleanup && { cleanup }) }
        }
      } catch (error) {
        console.warn('Handler failed to identify textarea:', error)
      }
    }
    return null
  }

  private handleDelayedValueInjection(overtype: OverTypeInstance): void {
    // GitHub sometimes injects textarea content after a delay
    // We need to trigger OverType to update its preview after such injections
    // https://github.com/diffplug/gitcasso/issues/46
    setTimeout(() => {
      overtype.updatePreview()
    }, 100)
    setTimeout(() => {
      overtype.updatePreview()
    }, 200)
    setTimeout(() => {
      overtype.updatePreview()
    }, 400)
    setTimeout(() => {
      overtype.updatePreview()
    }, 800)
  }

  getEnhancerCount(): number {
    return this.enhancers.size
  }
}

export class TextareaRegistry {
  private textareas = new Map<HTMLTextAreaElement, EnhancedTextarea>()
  private eventSender: (event: CommentEvent) => void = () => {}

  setCommentEventSender(sendEvent: (event: CommentEvent) => void): void {
    this.eventSender = sendEvent
  }

  private sendEvent(eventType: CommentEventType, enhanced: EnhancedTextarea): void {
    this.eventSender({
      draft: enhanced.textarea.value,
      spot: enhanced.spot,
      type: eventType,
    })
  }

  register<T extends CommentSpot>(enhanced: EnhancedTextarea<T>): void {
    this.textareas.set(enhanced.textarea, enhanced)
    enhanced.textarea.addEventListener('blur', () => {
      this.sendEvent('LOST_FOCUS', enhanced)
    })
    this.sendEvent('ENHANCED', enhanced)
  }

  unregisterDueToModification(textarea: HTMLTextAreaElement): void {
    const enhanced = this.textareas.get(textarea)
    if (enhanced) {
      if (enhanced.cleanup) {
        enhanced.cleanup()
      }
      this.sendEvent('DESTROYED', enhanced)
      this.textareas.delete(textarea)
    }
  }

  get(textarea: HTMLTextAreaElement): EnhancedTextarea | undefined {
    return this.textareas.get(textarea)
  }

  tabLostFocus(): void {
    for (const enhanced of this.textareas.values()) {
      this.sendEvent('LOST_FOCUS', enhanced)
    }
  }
}
