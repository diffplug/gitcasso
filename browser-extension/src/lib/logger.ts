import { CONFIG, type LogLevel } from './config'

/**
 * Simple logging utilities for the extension
 */

const prefix = `[${CONFIG.EXTENSION_NAME}]`

// No-op function for disabled logging
const noop = () => {}

// Log level hierarchy - index represents priority
const LOG_LEVEL_PRIORITY: Record<LogLevel, number> = {
  DEBUG: 0,
  ERROR: 3,
  INFO: 1,
  WARN: 2,
}

// Helper function to check if a log level is enabled
const shouldLog = (level: LogLevel): boolean => {
  // Don't log anything in production mode
  if (CONFIG.MODE === 'PROD') {
    return false
  }
  return LOG_LEVEL_PRIORITY[level] >= LOG_LEVEL_PRIORITY[CONFIG.LOG_LEVEL]
}

// Export simple logging functions
export const logger = {
  debug: shouldLog('DEBUG') ? console.log.bind(console, prefix) : noop,
  error: shouldLog('ERROR') ? console.error.bind(console, prefix) : noop,
  info: shouldLog('INFO') ? console.log.bind(console, prefix) : noop,
  time: shouldLog('INFO') ? console.time.bind(console) : noop,
  timeEnd: shouldLog('INFO') ? console.timeEnd.bind(console) : noop,
  warn: shouldLog('WARN') ? console.warn.bind(console, prefix) : noop,
}
