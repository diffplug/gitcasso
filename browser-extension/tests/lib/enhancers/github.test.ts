import { describe, expect, usingHar } from '../../har-fixture'

// must import fixture **first** for mocks, the `expect` keeps biome from changing sort-order
expect

import { EnhancerRegistry } from '../../../src/lib/registries'

describe('github', () => {
  usingHar('gh_pr').it('should create the correct spot object for gh_pr', async () => {
    const enhancers = new EnhancerRegistry()
    const textareas = document.querySelectorAll('textarea')
    expect(textareas.length).toBe(2)
    expect(enhancers.tryToEnhance(textareas[0]!)).toBeNull()
    expect(enhancers.tryToEnhance(textareas[1]!)?.spot).toMatchInlineSnapshot(`
      {
        "domain": "github.com",
        "number": 517,
        "slug": "diffplug/selfie",
        "type": "GH_PR_ADD_COMMENT",
        "unique_key": "github.com:diffplug/selfie:517",
      }
    `)
  })
  usingHar('gh_new_pr').it('should create the correct spot object for gh_new_pr', async () => {
    const enhancers = new EnhancerRegistry()
    const textareas = document.querySelectorAll('textarea')
    expect(textareas.length).toBe(2)
    expect(enhancers.tryToEnhance(textareas[0]!)?.spot).toMatchInlineSnapshot(`
          {
            "domain": "github.com",
            "slug": "diffplug/selfie/main...cavia-porcellus:selfie:main",
            "type": "GH_PR_NEW_COMMENT",
            "unique_key": "github.com:diffplug/selfie/main...cavia-porcellus:selfie:main",
          }
        `)
  })
  usingHar('gh_issue').it('should create the correct spot object for gh_issue', async () => {
    const enhancers = new EnhancerRegistry()
    const textareas = document.querySelectorAll('textarea')
    expect(textareas.length).toBe(1)
    expect(enhancers.tryToEnhance(textareas[0]!)?.spot).toMatchInlineSnapshot(`
      {
        "domain": "github.com",
        "number": 523,
        "slug": "diffplug/selfie",
        "type": "GH_ISSUE_ADD_COMMENT",
        "unique_key": "github.com:diffplug/selfie:523",
      }
    `)
  })
  usingHar(
    'gh_new_issue',
  ).it('should create the correct spot object for gh_new_issue', async () => {
    const enhancers = new EnhancerRegistry()
    const textareas = document.querySelectorAll('textarea')
    expect(textareas.length).toBe(1)
    expect(enhancers.tryToEnhance(textareas[0]!)?.spot).toMatchInlineSnapshot(`
          {
            "domain": "github.com",
            "slug": "diffplug/selfie",
            "type": "GH_ISSUE_NEW_COMMENT",
            "unique_key": "github.com:diffplug/selfie:new",
          }
        `)
  })
})
