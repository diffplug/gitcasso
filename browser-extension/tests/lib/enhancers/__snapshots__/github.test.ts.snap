// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GitHubHandler > should create correct GitHubContext spot for PR comment > github-pr-517-spot 1`] = `
{
  "domain": "github.com",
  "number": 517,
  "slug": "diffplug/selfie",
  "type": "GH_PR_ADD_COMMENT",
  "unique_key": "github:diffplug/selfie:pull:517",
}
`;
